webrender/src/device/gfx.rs

```diff
@@ -2137,8 +2124,45 @@ impl<B: hal::Backend> Device<B> {
         let renderer_name = "TODO renderer name".to_owned();
         let features = adapter.physical_device.features();
 
-        let memory_types = adapter.physical_device.memory_properties().memory_types;
-        println!("memory_types: {:?}", memory_types);
+        let memory_properties = adapter.physical_device.memory_properties();
+        let mut heaps = {
+            use std::cmp::min;
+            let types = memory_properties.
+                memory_types
+                .iter()
+                .map(|ref mt| {
+                    let config = HeapsConfig {
```

Comment:
I don't fully understand this initialization part. So we provide the configuration of every available memory type. That's ok. But why do we configure CPU-visible memory types with both Dynamic and Linear allocators? How is rendy supposed to figure out when to use linear over dynamic?

Reply:
Rendy picks based on usage.
For upload/download it prefers linear allocator.
For dynamic (uniforms, persistent mapping) it prefers dynamic allocator.